design_doc.md

```diff
@@ -0,0 +1,268 @@
+# XDGGS - Design document
+
+Xarrays extension for DGGS. Technical specifiactions.
+
+## Goals
+
+The goal of the `xddgs` library is to facilitate working with multiple Discrete Global Grid Systems (DGGSs) via a unified, high-level and user-friendly API that is deeply integrated with [Xarray](https://xarray.dev).
+
+Examples of common DGGS features that `xdggs` should provide or facilitate:
+
+- convert a DGGS from/to another grid (e.g., a DGGS, a latitude/longitude rectilinear grid, a raster grid, an unstructured mesh)
```

Comment:
In the discussion on Thursday, we pretty much agreed that this doesn't need to live in `xdggs` (though as part of the work on `xdggs` we need to ensure they exist). `xesmf` and `pyresample` seem like good candidates.

Edit: maybe that's what you meant with the section in the "non-goals"?

Reply:
Yes, what I mean here is that `xdggs` should provide the API for doing these operations (or at least some core API that should facilitate it). `xdggs` does not necessarily need to implement the logic: if there is a well-maintained 3rd party library that does that well, it is preferable to rely on it instead of reinventing the wheel.

About `xesmf` and `pyresample`, do they handle complex cell geometries (non-rectilinear) on the sphere?

Reply:
The 1st sentence of the "goals" section is particularly important: *"unified, high-level and user-friendly API that is deeply integrated with Xarray"*. It is all about the API so I expect that `xdggs` would mainly consist of adapter code that extend the Xarray API and calls routines from 3rd party libs.

Reply:
This is mostly a API design issue: does it make sense to provide a `interpolate` accessor method if we simply forward to another library? Oftentimes, using that library directly is more efficient, since at least in the case of `xesmf` it allows applying the same multiple times without a recomputation of weights. (I don't think there's a universal answer here, it depends on the library and how it works) 

> About `xesmf` and `pyresample`, do they handle complex cell geometries (non-rectilinear) on the sphere?

not sure, but the point is that if they don't we could (try to) extend them instead of implementing this ourselves. Note that those two libraries were examples, if there's a different library that does this already I wouldn't mind using that instead.

Reply:
> It is all about the API so I expect that xdggs would mainly consist of adapter code that extend the Xarray API and calls routines from 3rd party libs.

this would certainly be very convenient, and I didn't think of alternatives, yet (just to have more options to choose from). That will take me some time, though.

Reply:
> This is mostly a API design issue: does it make sense to provide a interpolate accessor method if we simply forward to another library? 

Perhaps not an interpolate accessor but conversion methods in a dggs accessor yes. I think it often makes sense since we're not only regridding but we're also conforming the output to some conventional (or standard?) DGGS representation used by `xdggs` (+ building/setting an index).

It is also about discoverability. We could certainly add tutorials showing how to regrid lat/lon 2D grids to DGGS by directly calling xesmf, pyreseample, etc., and then set the DGGS index. But I find it quite conventional to have some high-level API layer in xdggs to do the regridding and conversion all at once, since it is a very common operation.

Want to create a new DGGS Dataset from another Dataset? Just type `ds.dggs.from_<tab-completion>` and pick up the conversion method that is relevant for your dataset :).

>  at least in the case of xesmf it allows applying the same multiple times without a recomputation of weights. 

I think we could arrange the `xdggs` API so that it allows reusing interpolation weights. That doesn't seem a big issue to me.

> if there's a different library that does this already I wouldn't mind using that instead.

Totally agreed, we are on the same page!

Reply:
What I also tried to state in the doc (non-goals section): *"Likewise, there may be many ways of resampling a grid to a DGGS ; xdggs should support the most common methods but not try to support all of them."*

While I think we should really have conversion methods for converting lat/lon 2D grid to DGGS and vice-versa that handle 80% of the use-cases, it shouldn't prevent reusing directly xesmf, pyresample, etc. directly for the more advanced use cases.